runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/state/FlinkStateInternals.java

```diff
@@ -114,16 +127,27 @@ public K getKey() {
     return address.getSpec().bind(address.getId(), new FlinkStateBinder(namespace, context));
   }
 
-  public void clearBagStates(StateNamespace namespace, StateTag<? extends BagState> address)
-      throws Exception {
-    CoderTypeSerializer typeSerializer = new CoderTypeSerializer<>(VoidCoder.of());
-    flinkStateBackend.applyToAllKeys(
-        namespace.stringKey(),
-        StringSerializer.INSTANCE,
-        new ListStateDescriptor<>(address.getId(), typeSerializer),
-        (key, state) -> {
-          state.clear();
-        });
+  /**
+   * Allows to clear all state for the global watermark when the maximum watermark arrives. We do
+   * not clean up the global window state via timers because we are not guranteed to ever receive
+   * the final watermark which would lead to an unbounded number of keys and cleanup timers.
+   * Instead, the cleanup code below should be run when we finally receive the max watermark.
```

Comment:
I would remove "because we are not guaranteed to ever receive the final watermark" because that is misleading. The problem is that we accumulate too many timers, depending on the number of keys.